platform/linux-generic/arch/aarch64/odp_crypto_armv8.c

```diff
@@ -131,6 +158,179 @@ null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
+		ODP_DBG("ARM Crypto: AES GCM Encoding failed\n");
+		ret = -1;
+		goto err;
+	}
+
+	odp_packet_copy_from_mem(pkt, param->cipher_range.offset,
+				 param->cipher_range.length, local.buffer);
+
+	odp_packet_copy_from_mem(pkt, param->hash_result_offset,
+				 session->p.auth_digest_len, local.digest);
+
+err:
+	return ret < 0 ? ODP_CRYPTO_ALG_ERR_DATA_SIZE :
+			 ODP_CRYPTO_ALG_ERR_NONE;
+}
+
+static
+odp_crypto_alg_err_t aes_gcm_decrypt(odp_packet_t pkt,
+				     const odp_crypto_packet_op_param_t *param,
+				     odp_crypto_generic_session_t *session)
+{
+	armv8_cipher_state_t cs = {
+		.counter = {
+			.d = {0, 0}
+		}
+	};
+	uint8_t *iv_ptr;
+	uint8_t tag[16];
+	uint64_t plaintext_bit_length = param->cipher_range.length * 8;
+	uint64_t iv_bit_length = session->p.cipher_iv.length * 8;
+	uint64_t aad_bit_length = session->p.auth_aad_len * 8;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+	int ret = 0;
+
+	/* Fail early if cipher_range is too large */
+	if (in_len > ARM_CRYPTO_MAX_DATA_LENGTH) {
+		ODP_DBG("ARM Crypto: Packet size too large for requested operation\n");
+		ret = -1;
+		goto err;
+	}
+
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else
+		return ODP_CRYPTO_ALG_ERR_IV_INVALID;
+
+	cs.constants = &session->cc;
+
+	if (armv8_aes_gcm_set_counter(iv_ptr, iv_bit_length, &cs) != 0) {
+		ODP_DBG("ARM Crypto: Failure while setting nonce\n");
+		ret = -1;
+		goto err;
+	}
+
+	/* Copy current hash bytes to a buffer */
+	odp_packet_copy_to_mem(pkt, param->hash_result_offset,
+			       session->p.auth_digest_len, tag);
+	/* Clear hash bytes */
+	_odp_packet_set_data(pkt, param->hash_result_offset, 0,
+			     session->p.auth_digest_len);
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (in_len > seg_len) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (armv8_dec_aes_gcm_from_state(&cs,
+					 param->aad_ptr, aad_bit_length,
+					 data, plaintext_bit_length,
+					 tag,
+					 local.buffer) != 0) {
+		ODP_DBG("ARM Crypto: AES GCM Decoding failed\n");
+		ret = -1;
+		goto err;
+	}
+
+	odp_packet_copy_from_mem(pkt, param->cipher_range.offset,
+				 param->cipher_range.length, local.buffer);
+
+err:
+	return ret < 0 ? ODP_CRYPTO_ALG_ERR_ICV_CHECK :
+			 ODP_CRYPTO_ALG_ERR_NONE;
+}
+
+static int process_aes_gcm_param(odp_crypto_generic_session_t *session)
+{
+	/* Verify Key len is valid */
+	if (16 != session->p.cipher_key.length &&
+	    24 != session->p.cipher_key.length &&
+	    32 != session->p.cipher_key.length)
+		return -1;
+
+	/* Verify IV len is correct */
+	if (12 != session->p.cipher_iv.length)
+		return -1;
+
+	memcpy(session->cipher.key_data, session->p.cipher_key.data,
```

Comment:
It would be good to check here that cipher_key.length does not exceed the max length.

Reply:
I will add the check. Thanks!

Reply:
In the current code the memcpy is safe but the safety depends on the somewhat hidden assumption that ARM_CRYPTO_MAX_CIPHER_KEY_LENGTH is at least 32.

I am proposing an explicit check that cipher_key.length does not exceed ARM_CRYPTO_MAX_CIPHER_KEY_LENGTH, as otherwise the memcpy writes past the end of the array. This makes the code more robust (e.g. against changes of the value of the macro).

Reply:
Okay, I will add an assertion just before the memcpy operation. Thanks Janne!

Reply:
I think a runtime check would be better than assert here since the asserts are there only in debug builds and this is not fast path code that needs to be optimized. All the other checks before the added assert are also runtime checks and not asserts so this could just follow suit.

Reply:
Okay, sounds good. I will make the change. Thanks!